File: javascript/src/stacks/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
const isValid = function (s) {
  const bracketPairs = {
    ")": "(",
    "}": "{",
    "]": "[",
  };

  let stack = [];

  for (let i = 0; i < s.length; i++) {
    if (s[i] === "(" || s[i] === "{" || s[i] === "[") {
      stack.push(s[i]);
      continue;
    }

    const lastBracket = stack.pop();
    if (bracketPairs[s[i]] !== lastBracket) {
      return false;
    }
  }

  return stack.length === 0;
};

module.exports = isValid;

File: golang/stacks/0020-valid-parentheses.go
package stacks

func isValid(s string) bool {
	bracketPairs := map[rune]rune{
		')': '(',
		'}': '{',
		']': '[',
	}

	stack := []rune{}

	for _, r := range s {
		switch r {
		case '(', '{', '[':
			stack = append(stack, r)
		case ')', '}', ']':
			if len(stack) == 0 || bracketPairs[r] != stack[len(stack)-1] {
				return false
			}
			stack = stack[:len(stack)-1]
		}
	}

	return len(stack) == 0
}
